pdm/models/repositories.py

```diff
@@ -124,6 +131,30 @@ def find_candidates(
                 key=lambda c: c.version,
                 reverse=True,
             )
+
+            if not sorted_cans:
+                self.environment.project.core.ui.echo(
+                    "Could not find any matching candidates even when considering "
+                    "pre-releases",
+                    verbosity=termui.DETAIL,
+                    fg="yellow",
+                )
+
+        def print_candidates(title: str, candidates: List[Candidate]) -> None:
+            self.environment.project.core.ui.echo(
+                title, verbosity=termui.DETAIL, fg="blue"
+            )
+            with self.environment.project.core.ui.indent("  "):
+                for can in candidates:
+                    self.environment.project.core.ui.echo(
+                        can, verbosity=termui.DETAIL, fg="blue"
+                    )
+
+        if sorted_cans:
+            print_candidates("Found matching candidates:", sorted_cans)
+        else:
+            print_candidates("Found but non-matching candidates:", cans)
```

Comment:
This list may be super long for a popular package that has existed for years.

Reply:
How about limiting it to the top 10 (which might be arbitrary for `cans`)?